tasks/base.py

```diff
@@ -250,27 +245,32 @@ def _emit_queue_metrics(self):
             time_in_queue_timer.observe(delta.total_seconds())
 
     def run(self, *args, **kwargs):
-        task = get_current_task()
+        with self.task_full_runtime.time():  # Timer isn't tested
+            db_session = get_db_session()
 
-        if task and task.request:
-            if log_read_task_name() is not None or log_read_task_id() is not None:
-                log.warning(
-                    "There are multiple tasks concurrently writing to the task name at a given time"
-                )
-            log_set_task_name(task.name)
-            log_set_task_id(task.request.id)
+            log_context = LogContext(
+                repo_id=kwargs.get("repoid"),
+                owner_id=kwargs.get("ownerid"),
+                commit_sha=kwargs.get("commitid"),
+            )
 
-        self.task_run_counter.inc()
-        self._emit_queue_metrics()
+            task = get_current_task()
+            if task and task.request:
+                log_context.task_name = task.name
+                log_context.task_id = task.request.id
 
-        metric_context = MetricContext(
-            commit_sha=kwargs.get("commitid"),
-            repo_id=kwargs.get("repoid"),
-            owner_id=kwargs.get("ownerid"),
-        )
+            log_context.populate_from_sqlalchemy(db_session)
+            set_log_context(log_context)
```

Comment:
Would it make sense to call this `set_log_context()` every time inside the `populate_from_sqlalchemy` instead? I worry that people would subtly forget immediately add it after calling the `log_context.populate_from_sqlalchemy(db_session)` fn

Reply:
i can make `set_log_context()` call `populate_from_sqlalchemy()` internally

`populate_from_sqlalchemy()` sets/checks a `self._populated_from_db` flag that will skip the queries if it has already been run. i figure that when the task begins you already know the relevant owner, repo, and commit, if they exist. so you can construct the `LogContext` with what you have, fill in any missing info you can from the database, and then lock it in.

Reply:
oh actually i don't know if i want to do that

`set_log_context()` is called after `wrap_up_dbsession(dbsession)` so i don't know if it is safe to call `populate_from_sqlalchemy()` at that point. it would be a no-op because it's called with an empty `LogContext`, but that feels sketchy to rely on

Reply:
Does this mean you need to run `set_log_context()` every time after `log_context.populate_from_sqlalchemy(db_session)` is called?

Reply:
if you want to update the contextvar yes, but the db fetch probably doesn't need to happen anywhere except right here